### ISSUE TITLE
Float64List.setAll issue on dart2js and dartdevk

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#43227]
The following [test](https://github.com/dart-lang/co19/blob/master/LibTest/typed_data/Float64List/setAll_A04_t01.dart) works well in VM but fails on dart2js and dartdevk
```dart

import "dart:typed_data";
import "../../../Utils/expect.dart";

main() {
  var l = new Float64List.fromList(
      [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]);

  l.setAll(4, l.getRange(1, 3));
  Expect.listEquals([0.0, 1.0, 2.0, 3.0, 1.0, 2.0, 6.0, 7.0, 8.0, 9.0], l);

  l.setAll(2, l.getRange(1, 4));
  Expect.listEquals([0.0, 1.0, 1.0, 2.0, 3.0, 2.0, 6.0, 7.0, 8.0, 9.0], l); // deepEquals([0, 1, 1, 2, 3, 2, 6, 7, 8, 9], [0, 1, 1, 1, 1, 2, 6, 7, 8, 9], ) fails

  l.setAll(0, l.getRange(2, 7));
  Expect.listEquals([1.0, 2.0, 3.0, 2.0, 6.0, 2.0, 6.0, 7.0, 8.0, 9.0], l);
}
```
`Dart SDK version: 2.10.0-56.0.dev (dev) (Mon Aug 24 09:17:19 2020 -0700) on "windows_x64"`

**Comment by lrhn** [comment]
Seems JS compiled `setAll` does not handle copying a range into an overlapping region safely.

I'm not sure it *must*, we can't see directly that `l.getRange` is part of `l`, it's just some iterable.

The only operation guaranteed to recognize copying into an overlapping region of the same list is `setRange`.

**Comment by rakudrama** [comment]
TL;DR: I am supportive of making the implementations closer, but I thing that means changing the VM version.

The VM's `_DoubleListMixin` in [typed_data_patch.dart](https://github.com/dart-lang/sdk/blob/master/sdk/lib/_internal/vm/lib/typed_data_patch.dart#L733) does this:
```dart
  void setAll(int index, Iterable<double> iterable) {
    final end = iterable.length + index;
    setRange(index, end, iterable);
  }
```
DDC and dart2js fall back on `ListMixin`:
```dart
  void setAll(int index, Iterable<E> iterable) {
    if (iterable is List) {
      setRange(index, index + iterable.length, iterable);
    } else {
      for (E element in iterable) {
        this[index++] = element;
      }
    }
  }
```
The result of `getRange` is _not_ a `List`, so only on the VM does `setAll` flow through to `setRange`.

However, I think the VM version is problematic. It contains the classic double-iteration antipattern.
Consider:
```dart
import 'dart:typed_data';

main() {
  var a = Float64List(10);
  a.setAll(0, [1.1, -2.2, 3.3].where(foo));
}

bool foo(double x) {
  print(x);
  return x >= 0;
}
```
VM prints the elements twice, but dart2js and DDC print them once.
I think this is a more serious problem that one one reported.

We could make the VM implementation the argument (e.g. `is EfficientLengthIterable`), but this test should be in agreement with the implementations of `setRange`, otherwise we end up copying the input:
```dart
    if (from is List<int>) {
      otherList = from;
      otherStart = skipCount;
    } else {
      otherList = from.skip(skipCount).toList(growable: false);
      otherStart = 0;
    }
```

Simply removing the VM implementations of `setAll` would cause the code to use `setRange` for lists only via `ListMixin.setAll`.
